WikiWatershed/model-my-watershed
# Fix JavaScript tests

## Overview

Partially reverts 91e7c54 and selectively applies changes. The NPM shrinkwrap update had been overly aggressive and changed libraries unrelated to Clipboard copying. Specifically, [`updateModificationHash`](https://github.com/WikiWatershed/model-my-watershed/blob/7c81929273fcb5878999092cc3b1fcb29abff36e/src/mmw/js/src/modeling/models.js#L846) was generating a different hash because the rounding of numbers had subtly changed:

```diff
--- 
+++ 
@@ -1,13 +1,13 @@
 [
   [
     {
-      "area": 44.42257988387458
+      "area": 44.422579883894436
     },
     {
-      "effectiveArea": 44.42257988387458
+      "effectiveArea": 44.422579883894436
     },
     {
       "effectiveShape": [
         {
           "geometry": [
             {
```

which led to a different hash being generated in Jenkins http://civicci01.internal.azavea.com/job/model-my-watershed-develop/750/console:

```
not ok 97 Firefox 54.0 - Modeling Models ScenarioModel #updateModificationHash generates and sets modification_hash based on the scenario's modifications
    ---
        message: >
            expected 'ca1872c7647c7cb9a7f0a00b683407df' to equal '582b6178186440159bd4ea25f0260892'
    ...
```

By hand picking the updated libraries and keeping the rest the same, we ensure that the calculation integrity is maintained.

The diff might not be readable in GitHub, but if we look at changes in this file from before it was updated and this branch, we see it is much more specific and manageable:

```bash
git diff 2902441..63434d6 src/mmw/npm-shrinkwrap.json
```

```diff
diff --git a/src/mmw/npm-shrinkwrap.json b/src/mmw/npm-shrinkwrap.json
index 3eaf69e..7e42a7e 100644
--- a/src/mmw/npm-shrinkwrap.json
+++ b/src/mmw/npm-shrinkwrap.json
@@ -1052,6 +1052,35 @@
         }
       }
     },
+    "clipboard": {
+      "version": "1.7.1",
+      "from": "clipboard@*",
+      "resolved": "https://registry.npmjs.org/clipboard/-/clipboard-1.7.1.tgz",
+      "dependencies": {
+        "good-listener": {
+          "version": "1.2.2",
+          "from": "good-listener@>=1.2.2 <2.0.0",
+          "resolved": "https://registry.npmjs.org/good-listener/-/good-listener-1.2.2.tgz",
+          "dependencies": {
+            "delegate": {
+              "version": "3.1.3",
+              "from": "delegate@>=3.1.2 <4.0.0",
+              "resolved": "https://registry.npmjs.org/delegate/-/delegate-3.1.3.tgz"
+            }
+          }
+        },
+        "select": {
+          "version": "1.1.2",
+          "from": "select@>=1.1.2 <2.0.0",
+          "resolved": "https://registry.npmjs.org/select/-/select-1.1.2.tgz"
+        },
+        "tiny-emitter": {
+          "version": "2.0.1",
+          "from": "tiny-emitter@>=2.0.0 <3.0.0",
+          "resolved": "https://registry.npmjs.org/tiny-emitter/-/tiny-emitter-2.0.1.tgz"
+        }
+      }
+    },
     "d3": {
       "version": "3.5.5",
       "from": "https://registry.npmjs.org/d3/-/d3-3.5.5.tgz",
@@ -4597,11 +4626,6 @@
           "resolved": "https://registry.npmjs.org/xtend/-/xtend-4.0.0.tgz"
         }
       }
-    },
-    "zeroclipboard": {
-      "version": "2.2.0",
-      "from": "https://registry.npmjs.org/zeroclipboard/-/zeroclipboard-2.2.0.tgz",
-      "resolved": "https://registry.npmjs.org/zeroclipboard/-/zeroclipboard-2.2.0.tgz"
     }
   }
 }
```

### Notes

There is a further failing test in Testem in Chrome that doesn't fail in other environments. That has not been fixed here.

## Testing Instructions

 * Check out this branch
 * Delete your `node_modules` folder: `vagrant ssh app -c 'cd /opt/app && rm -rf node_modules'`
 * Install new npm modules: `./scripts/npm.sh install`
 * Bundle: `./scripts/bundle.sh --tests --vendor`
 * Run Testem `./scripts/testem.sh` and ensure that [this test](http://localhost:7357/test.html?grep=Modeling%20Models%20ScenarioModel%20%23updateModificationHash%20generates%20and%20sets%20modification_hash%20based%20on%20the%20scenario%27s%20modifications) passes
 * Run `./scripts/test.sh` and ensure that everything passes
